File: 1130-minimum-cost-tree-from-leaf-values/1130-minimum-cost-tree-from-leaf-values.cpp
class Solution {
public:
    vector<vector<pair<int,int>>>dp;
    pair<int,int> recur(vector<int>&arr,int l,int r){
        if(l==r){
            return {arr[l],0};
        }
        if(l>r)return {0,1000000};
        if(dp[l][r].second!=-1)return dp[l][r];
        pair<int,int>ans={0,1000000};
        for(int i=l;i<r;i++){
            pair<int,int>p_left=recur(arr,l,i);
            pair<int,int>p_right=recur(arr,i+1,r);
            int total=p_left.second+p_right.second+(p_left.first*p_right.first);
            if(total<ans.second){
                ans.second=total;
                ans.first=max(p_left.first,p_right.first);
            }
        }
        return dp[l][r]= ans;
    }
    
    int mctFromLeafValues(vector<int>& arr) {
        dp.resize(arr.size(),vector<pair<int,int>>(arr.size(),{-1,-1}));
        return recur(arr,0,arr.size()-1).second;
    }
};

File: 1130-minimum-cost-tree-from-leaf-values/1130-minimum-cost-tree-from-leaf-values.java
class Solution {
    
    class pair{
        int maxi;
        int sum;
        pair(int maxi,int sum){
            this.maxi=maxi;
            this.sum=sum;
        }
    }
    
    public pair memo(int l,int r,int [] arr,pair dp[][]){
        if(l==r)return new pair(arr[l],0);
        if(dp[l][r]!=null)return dp[l][r];
        pair ans=new pair(0,1000000);
        for(int i=l;i<r;i++){
            pair p_left=memo(l,i,arr,dp);
            pair p_right=memo(i+1,r,arr,dp);
            int total=p_left.sum+p_right.sum+(p_left.maxi*p_right.maxi);
            if(total<ans.sum){
                ans.sum=total;
                ans.maxi=Math.max(p_left.maxi,p_right.maxi);
            }
        }
        return dp[l][r]=new pair(ans.maxi,ans.sum);
        
    }
    
    
    public int mctFromLeafValues(int[] arr) {
        return memo(0,arr.length-1,arr,new pair[arr.length][arr.length]).sum;
    }
}